### PR TITLE
fix: scan encoded pickle metadata within byte budget

### PR DESCRIPTION
**Pull request by mldangelo-oai**
## Summary

Benign PyTorch metadata can exceed the old 64-token limit and make an otherwise clean scan exit 2. Scan small encoded tokens within a 4,096-token limit and the existing decoded-byte budget. Retain the 64-token bound when oversized unlabelled candidates are skipped.

Inspect labelled oversized tokens within the byte budget, report incomplete coverage when a limit is reached, and preserve late execution-pattern detection within a bounded follow-up budget.

## Campaign Evidence

Paired static replay on 2026-09-09 used the same artifact, Python dependency runtime and scan options (`--no-whitelist --no-cache --max-size 8GB --timeout 5400`). Imported source paths, clean worktrees and artifact hashes were verified before and after scanning.

- Model: `amd/realesrgan-x4plus`, HF revision `9af1b12e679374d2f0b0ccdb2488cee92e2525e1`.
- Artifact: `RealESRGAN_x4plus.pth`, 67,040,989 bytes, SHA-256 `4fa0d38905f75ac06eb49a7951b426670021be3018265fd191d2125df9d682f1`.
- Main `9631527b3e81e0458fe9e0242c3178af406bfa5f`: exit 2, `success=false`, S902 `pickle_encoded_text_scan_limit_exceeded` on `archive/data.pkl`.
- PR `38d0479d2d0d8b97ccdd3c6bab90311a06e46266`: exit 0, `success=true`, no issues or failed checks.
- Both runs recognize one PyTorch ZIP artifact and report 322,652 scanned bytes. This is a regression replay, not an additional completed-main campaign model.

## Validation

- 41 focused pickle/PyTorch ZIP tests passed using `-k 'encoded or base64 or seedless or shifted'`.
- 48 bounded static comparisons cover binary/printable and labelled/unlabelled tokens, shared byte budgets, benign controls and late execution-pattern detection. Both previously reported gaps now produce S604 or fail closed with S902.
- Full `tests/scanners/test_pickle_scanner.py` and `tests/scanners/test_pytorch_zip_scanner.py` runs passed on the coordinator.
- Current-head CI lint, formatting and type checks passed; the full CI result remains tracked separately on this PR.

## Local Limitations

- The canonical local mypy command is blocked on this 3.12-only coordinator because project config targets Python 3.10 while the local NumPy 2.5 stubs use Python 3.12 `type` syntax. A Python 3.12 full mypy run then reported one unrelated unreachable TensorFlow import in `tests/scanners/test_weight_distribution_scanner.py`.
- The full local non-slow pytest-xdist gate was inconclusive: it recorded failures outside the touched tests, then spent several minutes draining long `tests/detectors/test_cve_detection.py_cases/test_pickle_binunicode8_setitem.py` parameters without producing a final traceback before interruption. The touched tests and relevant PyTorch ZIP suite pass under xdist locally; repository CI should be treated as the full-gate authority.

### CODE REVIEW

**Comment by mldangelo-oai** [comment]
@codex review

@codex security review

File: modelaudit/scanners/pickle_scanner.py [hunk 1]
```diff
@@ -4705,25 +4747,74 @@ def _scan_encoded_text_indicators(
     ) -> None:
         for encoding, token_pattern in (("base64", _BASE64_TOKEN_RE), ("hex", _HEX_TOKEN_RE)):
             seen_tokens: set[bytes] = set()
+            analyzed_tokens: set[bytes] = set()
+            seen_prefixed_tokens: set[bytes] = set()
             decoded_budget = _MAX_RAW_ENCODED_BYTES
-            token_count = 0
+            skipped_decoded_budget = _MAX_RAW_ENCODED_BYTES
```
[the remark this answers is not in the document]
Comment (mldangelo-oai) [reply modelaudit/scanners/pickle_scanner.py]: Confirmed independently on e0b2de2edbb4cf611cbd204ceb5fee3a80715330 using generated inert text only. For both Base64 and hex, 600,000 bytes of decoded small tokens plus a 600,000-byte skipped token report success with no encoded-text coverage check. Reversing token order has the same result. The combined 1,200,000 bytes exceed the declared 1,048,576-byte limit; unchanged main 9631527b3e81e0458fe9e0242c3178af406bfa5f reports incomplete coverage for these cases.

The under-budget controls preserve the useful token-cap removal, so reverting the whole fix is unnecessary. Please account for analyzed plus omitted candidate bytes in the coverage limit, in both processing orders, while preserving the actual decode-work bound and the existing late-detection controls. Keep reported analyzed bytes separate from bytes only accounted for. The merge hold remains until this is fixed and validated.

File: modelaudit/scanners/pickle_scanner.py [hunk 2]
```diff
@@ -4705,26 +4748,98 @@ def _scan_encoded_text_indicators(
...
                 has_seed = (
                     any(seed in token for seed in _BASE64_CODE_EXECUTION_SEEDS)
                     if encoding == "base64"
                     else _hex_token_has_execution_seed(token)
                 )
+                decoded_size = _encoded_token_decoded_size(encoding, token)
                 if len(token) > _MAX_RAW_ENCODED_TOKEN_WITHOUT_SEED_BYTES and not has_seed:
-                    continue
+                    if prefix_encoding is None or decoded_size is None:
+                        continue
```
[the remark this answers is not in the document]
Comment (mldangelo-oai) [reply modelaudit/scanners/pickle_scanner.py]: Confirmed on 43c2ed3672f2c81f7c47bcbdeb494ff3afe44401 with static, inert fixtures. The previous eight shared-budget cases now pass: 1,200,000 combined decoded/skipped bytes fail closed in both token orders and both encodings, while the 750,000-byte controls stay clean.

The remaining unprefixed gap reproduces for both Base64 and hex: 65 distinct 20,000-byte decoded values (1,300,000 bytes total) produce success with no S902 on this head; main 9631527 fails closed. Eight such values remain clean on both. Charge valid oversized candidates independently of labels while keeping actual decoding bounded.

I also corroborated the separate late-token review: after eight prefixed 131,072-byte values fill the coverage budget, the 40-byte static Base64 execution-pattern control produces S604, but its 20,000-byte version is skipped and leaves only S902. These checks only scan inert text; no payload was executed.

**Comment by mldangelo-oai** [comment]
@codex review
@codex security review

File: modelaudit/scanners/pickle_scanner.py [hunk 3]
```diff
@@ -4705,45 +4803,143 @@ def _scan_encoded_text_indicators(
...
-                    elif len(token) % 2 == 0:
-                        decoded = binascii.unhexlify(token)
-                    else:
+                    if decoded_size is None:
                         continue
-                except (binascii.Error, ValueError):
+                    if prefix_encoding is None and not _unprefixed_seedless_encoded_token_looks_textual(
+                        encoding, token
+                    ):
+                        continue
```
[the remark this answers is not in the document]
Comment (mldangelo-oai) [reply modelaudit/scanners/pickle_scanner.py]: Confirmed both current-head P1 reviews on 6c633dc54ee2bba847c3aeffe0d896b6ec61047b with bounded, inert static detector calls:

- 65 distinct unprefixed Base64 tokens decoding to 20,000 bytes each, padded with NUL bytes, return success with no S604 or S902 despite totaling 1,300,000 decoded bytes. The last token can contain an alignment-shifted execution-pattern literal. Main 9631527 fails closed. The printable-ratio filter still excludes valid encoded candidates from coverage accounting.
- The separate under-budget case also reproduces: 65 tokens of 3,100 decoded bytes total only 201,500 bytes. A shifted execution-pattern literal in the last token returns success with no S604 or S902, with or without an encoded: prefix and with printable or binary padding. Main fails closed at its token bound. Charging bytes without inspecting the skipped candidates does not close this gap.

The previous shared-budget cases remain fixed, the 1,300,000-byte printable unprefixed cases now fail closed in both Base64 and hex, and the earlier late 20,000-byte Base64 positive now produces S604. Eight-token benign controls remain clean. All execution-pattern literals were only scanned as bytes, never evaluated or executed.

Prefer one bounded decode/accounting path that inspects valid candidates and explicitly reports incomplete coverage for work it cannot cover. More independent printability, sampling, and overflow exceptions make these contracts harder to preserve. Keep these two current-head reviews unresolved until their positive and benign controls are covered.

**Comment by mldangelo-oai** [comment]
@codex review
@codex security review

**Review by mldangelo-oai — COMMENTED** [review]
Reviewed commit `38d0479d2d0d8b97ccdd3c6bab90311a06e46266` against `9631527b3e81e0458fe9e0242c3178af406bfa5f`. Both previously reported detection gaps are fixed. I found no new blocking source issues.

Validation: 41 focused pickle/PyTorch ZIP tests passed, plus 48 bounded static comparisons covering binary and printable tokens, labelled and unlabelled inputs, shared byte limits, benign controls, and late encoded detection. The previous bypass cases now produce S604 or fail closed with S902.

The requested pinned RealESRGAN replay is now complete. The same artifact at HF revision `9af1b12e679374d2f0b0ccdb2488cee92e2525e1` (SHA-256 `4fa0d38905f75ac06eb49a7951b426670021be3018265fd191d2125df9d682f1`) exits 2 with S902 on exact main and 0 with no issues on this PR head. Both runs use the same dependency runtime and scan options, recognize one PyTorch ZIP artifact, and report 322,652 scanned bytes. The PR description now reflects this replay and the final token limits.

Current-head code and security reviews completed without findings. Required CI and independent approval remain pending.
